src/server/notifications/WebSocketSubscription2021/WebSocket2021Listener.ts

```diff
@@ -0,0 +1,56 @@
+import type { IncomingMessage } from 'http';
+import type { WebSocket } from 'ws';
+import type { InteractionRoute } from '../../../identity/interaction/routing/InteractionRoute';
+import { getLoggerFor } from '../../../logging/LogUtil';
+import { WebSocketServerListener } from '../../WebSocketServerListener';
+import type { SubscriptionStorage } from '../SubscriptionStorage';
+import type { WebSocket2021Handler } from './WebSocket2021Handler';
+
+/**
+ * Listens for WebSocket connections and verifies if they are valid WebSocketSubscription2021 connections,
+ * in which case its {@link WebSocket2021Handler} will be alerted.
+ */
+export class WebSocket2021Listener extends WebSocketServerListener {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly storage: SubscriptionStorage;
+  private readonly handler: WebSocket2021Handler;
+  private readonly path: string;
+
+  public constructor(storage: SubscriptionStorage, handler: WebSocket2021Handler, route: InteractionRoute) {
+    super();
+    this.storage = storage;
+    this.handler = handler;
+    this.path = new URL(route.getPath()).pathname;
+  }
+
+  protected async handleConnection(webSocket: WebSocket, upgradeRequest: IncomingMessage): Promise<void> {
+    // Base doesn't matter since we just want the path and query parameter
+    const url = new URL(upgradeRequest.url ?? '', 'http://example.com');
```

Comment:
then perhaps for safety `{ pathname, searchParams } = `, such that we can never access the invalid URL.